### ISSUE TITLE
Run-time error

### ISSUE THREAD

**Issue by epoberezkin** [MatAtBread/nodent#18]
This code:

```
validate = async function(data, dataPath) {
  'use strict';
  var vErrors = null;
  var errors = 0;
  if ((data && typeof data === "object" && !Array.isArray(data))) {
    var errs__0 = errors;
    var valid1 = true;
    for (var key0 in data) {
      var data1 = data[key0];
      var errs_1 = errors;
      if (typeof data1 === "string") {
        var schema1 = key0;
        var format1 = formats[schema1];
        var isObject1 = typeof format1 == 'object' && !(format1 instanceof RegExp) && format1.validate;
        if (isObject1) {
          var async1 = format1.async;
          format1 = format1.validate;
        }
        if ((schema1 !== undefined && typeof schema1 != 'string') || (format1 && !(typeof format1 == 'function' ? (async1 ? await format1(data1) : format1(data1)) : format1.test(data1)))) {
          throw new ValidationError([{
            keyword: 'format',
            dataPath: (dataPath || '') + '[\'' + key0 + '\']',
            schemaPath: "#/additionalProperties/format",
            params: {
              format: schema1
            },
            message: 'should match format "' + schema1 + '"'
          }]);
        }
      } else {
        throw new ValidationError([{
          keyword: 'type',
          dataPath: (dataPath || '') + '[\'' + key0 + '\']',
          schemaPath: "#/additionalProperties/type",
          params: {
            type: 'string'
          },
          message: 'should be string'
        }]);
      }
      var valid1 = errors === errs_1;
      if (!valid1) break;
    }
  }
  if (errors === 0) return true;
  else throw new ValidationError(vErrors);
};
```

Transpiles to:

```
validate = function (data, dataPath) {
    return new Promise((function ($return, $error) {
        var vErrors, errors, errs__0, valid1, data1, errs_1, schema1, format1, isObject1, async1;
        'use strict';
        vErrors = null;
        errors = 0;
        function $IfStatement_3() {
            if (errors === 0) 
                return $return(true);
             else 
                return $error(new ValidationError(vErrors));
            return $return();
        }
        
        if (data && typeof data === "object" && !Array.isArray(data)) {
            errs__0 = errors;
            valid1 = true;
            var $in_1 = [];
            for (var key0 in data) 
                $in_1.push(key0);
            return (function $WhileStatement_2_loop($WhileStatement_2_exit, $error) {
                function $WhileStatement_2_next() {
                    return $WhileStatement_2_loop($WhileStatement_2_exit, $error);
                }
                
                if ($in_1.length) {
                    key0 = $in_1.shift();
                    data1 = data[key0];
                    errs_1 = errors;
                    function $IfStatement_5() {
                        valid1 = errors === errs_1;
                        if (!valid1) 
                            return void $WhileStatement_2_exit();
                        return void $WhileStatement_2_next.call(this);
                    }
                    
                    if (typeof data1 === "string") {
                        schema1 = key0;
                        format1 = formats[schema1];
                        isObject1 = typeof format1 == 'object' && !(format1 instanceof RegExp) && format1.validate;
                        if (isObject1) {
                            async1 = format1.async;
                            format1 = format1.validate;
                        }
                        return (function $logicalOr_8($await_9, $then) {
                            if ($then && !$await_9) 
                                return $logicalOr_8.$asyncbind(this)($await_7);
                            return (function $logicalAnd_6($await_7, $then) {
                                return (function $condOp($test, $then) {
                                    if ($then) {
                                        if ($test) {
                                            return format1(data1).then($condOp.$asyncbind(this), $error);
                                        } else 
                                            return $condOp.$asyncbind(this)(format1(data1));
                                    }
                                    if ($then && $await_7) 
                                        return $logicalAnd_6.$asyncbind(this)(!(typeof format1 == 'function' ? $test : format1.test(data1)));
                                    if ($await_9) {
                                        return $error(new ValidationError([{
                                            keyword: 'format',
                                            dataPath: (dataPath || '') + '[\'' + key0 + '\']',
                                            schemaPath: "#/additionalProperties/format",
                                            params: {
                                                format: schema1
                                            },
                                            message: 'should match format "' + schema1 + '"'
                                        }]));
                                    }
                                    return $IfStatement_5.call(this);
                                }).$asyncbind(this)(async1, 1);
                            }).$asyncbind(this)(format1, 1);
                        }).$asyncbind(this)(schema1 !== undefined && typeof schema1 != 'string', 1);
                    } else {
                        return $error(new ValidationError([{
                            keyword: 'type',
                            dataPath: (dataPath || '') + '[\'' + key0 + '\']',
                            schemaPath: "#/additionalProperties/type",
                            params: {
                                type: 'string'
                            },
                            message: 'should be string'
                        }]));
                    }
                    return $IfStatement_5.call(this);
                } else 
                    return void $WhileStatement_2_exit();
            }
            ).$asyncbind(this).then((function ($await_11) {
                return $IfStatement_3.call(this);
            }).$asyncbind(this, $error), $error);
        }
        return $IfStatement_3.call(this);
    }).$asyncbind(this));
};
```

and throws an exception at run time:

```
ReferenceError: $await_7 is not defined
      at $logicalOr_8 (eval at localCompile (lib/compile/index.js:82:12), <anonymous>:47:70)
      at $WhileStatement_2_loop (eval at localCompile (lib/compile/index.js:82:12), <anonymous>:72:44)
```

when `validate({ english_word: 'tomorrow' })` is called.

Not sure if it helps, but `format1('tomorrow')` returns promise that resolves to true.

**Comment by epoberezkin** [comment]
btw, the same function works ok when it uses generators (both with native and when transpiled with regenerator)

**Comment by epoberezkin** [comment]
It passed 10 other tests though :)

**Comment by matAtWork** [comment]
That's a good catch. I only (relatively) recently implemented transforms for `for (x in obj)` - looks like I might have missed an edge case. Can you point me at the test case that's failing and when I get some time I'll try and cut the code down to the structure that's failing

**Comment by matAtWork** [comment]
Out of interest, does it work with {generators:true} rather than Promises in nodent too?

**Comment by epoberezkin** [comment]
[This test](https://github.com/epoberezkin/ajv/blob/async/spec/async_validate.spec.js#L137) fails.

To run it with nodent you need to comment lines 30-37 in this file and uncomment lines 34 (and/or 35).

It fails on the first data sample already (line 155).

I passes with `{generators:true}` but it's not different from passing with native generators.

**Comment by matAtWork** [comment]
It's failing on the nasty conditional if expression - see http://nodent.mailed.me.uk/#async%20function%20validate(x%2Cy%2Cz)%20%7B%0A%20%20%20%20if%20(x%20!%3D%3D%20undefined%20%26%26%20typeof%20x%20!%3D%20'string'%20%7C%7C%20y%20%26%26%20!(typeof%20y%20%3D%3D%20'function'%20%3F%20z%20%3F%20await%20y(x)%20%3A%20y(x)%20%3A%20z.test(x)))%20%7B%0A%20%20%20%20%20%20%20%20throw%20new%20Error()%3B%0A%20%20%20%20%7D%0A%7D%0A%0Aawait%20validate(%22a%22%2C%22b%22%2C%22c%22)%20%3B

I'll try and pick apart what I've done wrong. In the meanwhile, I have a suspicion that simplifying the expression would fix it, if you want to try a workaround until I've fixed it.

**Comment by epoberezkin** [comment]
In my case it's `(x !== undefined && typeof x != 'string') || (y && !(typeof y == 'function' ? z ? await y(x) : y(x) : z.test(x)))`, so it's not ambiguous.

But it also fails.

I'll try the workaround, but there many other long conditionals in some cases, some are even generated in loops... Although there won't be any await in those places.

**Comment by epoberezkin** [comment]
Actually, I'd rather make more tests ;)

**Comment by matAtWork** [comment]
I've distilled the core error, its:
```
async function y() { return "XXX" }
console.log("3",0 || (y && await y()));
console.log("done");
```
The issue is in the transformation of `expr || (expr && await expr())`. It's a bug and I'll track it down and re-release

**Comment by matAtWork** [comment]
Fixed in 2.3.11 - thanks for the input
